### DISCUSSION TITLE
tutorials/authorize-webapp broken: "error": "invalid_grant"

### DISCUSSION THREAD

**Discussion by kevinrss01** [iTwin/itwinjs-core#5167] [General]
Hi,
I'm trying to run the tutorial (https://developer.bentley.com/tutorials/authorize-webapp/#2-get-an-access-token) to authorize my webapp (created in Trial mode), but on step 2 I always get this error message:
POST https://ims.bentley.com/connect/token --->
{
    "error_description": "Authorization code is invalid or expired.",
    "error": "invalid_grant"
}
Can someone confirm this nice tutorial page works for their webapp?
Thanks

**Comment by pratikshan85** [comment]
@kevinrss01 yes, this tutorial works in webapp.
You need to pass the correct query parameter.

query parameters:

**client_id** - This is the unique identifier for your application. Displayed on the application details page as Client ID (e.g. webapp-example).
**redirect_uri** - Specifies where you want to go back after successful authorization (e.g. http://localhost:3000/redirect).
**scope - List** of accesses granted to the application. Displayed on the application details page as Scopes (e.g. users:read).
**response_type** - For this flow code is the only option and it indicates that you will get code as a query parameter when the user will come back from the authorization server.
So in the end it should look something like this: (https://ims.bentley.com/connect/authorize?client_id=CLIENT_ID&redirect_uri=http%3A%2F%2Flocalhost%3A3000&scope=users%3Aread+offline_access&response_type=code)

You can also send us your query parameter for which it is not working.
If it is confidential then you can send us the details at: pratikshan.ghosh@bentley.com or arnab.gogoi@bentley.com

**Reply by kevinrss01** [reply]
I think there is a misunderstanding. 
I am talking about the request to "https://ims.bentley.com/connect/token" which asks as parameters: 
- **client_id** : myClientID
- **client_secret** : MyClientSecret
- **grant_type** : authorization_code
- **code** : code returned from https://ims.bentley.com/connect/authorize?{params}
- **redirect_uri** : http%3A%2F%2Flocalhost%3A5000%2Fauth%2Fauthorizationserver

**Reply by pratikshan85** [reply]
Thanks @kevinrss01 for your update.
I got the error & your API call.

Can you please check and try the same process, It should work now.
If still it is not working, please provide us with all the details.

client_id : myClientID - > exact value
client_secret : MyClientSecret -> exact value
grant_type : authorization_code -> exact value 
code : code returned from [https://ims.bentley.com/connect/authorize?{params}](https://ims.bentley.com/connect/authorize?%7Bparams%7D)
redirect_uri : http%3A%2F%2Flocalhost%3A5000%2Fauth%2Fauthorizationserver

**Reply by kevinrss01** [reply]
Thanks @pratikshan85, it works on your website, but on my end, I'm still getting the same error :
"Error: {"error_description":"Authorization code is invalid or expired.","error":"invalid_grant"}"

Here is an example of my request code : 

```
interface TokenResponse {
      client_id: string;
      client_secret: string;
      grant_type: string;
      code: string;
      redirect_uri: string;
    }

    const data: TokenResponse = {
      client_id: "MY_CLIENT_SECRET",
      client_secret:
        "MY_CLIENT_SECRET",
      grant_type: "authorization_code",
      code: typeof req.query.code === "string" ? req.query.code : "",
      redirect_uri:
        "http%3A%2F%2Flocalhost%3A5000%2Fauth%2Fauthorizationserver",
    };

    const params = new URLSearchParams();
    for (const [key, value] of Object.entries(data)) {
      params.append(key, value);
    }

    axios
      .post<TokenResponse>(
        "https://ims.bentley.com/connect/token",
        params.toString(),
        {
          headers: {
            "Content-Type": "application/x-www-form-urlencoded",
          },
        }
      )
      .then((response) => {
        console.log(response.data);
      })
      .catch((error) => {
        console.log("Error: " + JSON.stringify(error.response.data));
      });
```
      Maybe the issue come from that I use localhost ?

**Reply by pratikshan85** [reply]
@kevinrss01  can you please take a look on the attached example application below which generates access_token?
[app.zip](https://github.com/iTwin/itwinjs-core/files/10954291/app.zip)

Above application does the same job as you are doing above.
Please let me know if you need any further details.
